rllib/policy/tf_mixins.py

```diff
@@ -196,38 +198,43 @@ def set_state(self, state: PolicyState) -> None:
 
 
 class TargetNetworkMixin:
-    """Assign the `update_target` method to the SimpleQTFPolicy
+    """Assign the `update_target` method to the policy.
 
     The function is called every `target_network_update_freq` steps by the
     master learner.
     """
 
-    def __init__(
-        self,
-        obs_space: gym.spaces.Space,
-        action_space: gym.spaces.Space,
-        config: AlgorithmConfigDict,
-    ):
-        trainable_q_func_vars = self.model.trainable_variables()
-        trainable_target_q_func_vars = self.target_model.trainable_variables()
+    def __init__(self, update_target_in_init=True):
+
+        model_vars = self.model.trainable_variables()
+        target_model_vars = self.target_model.trainable_variables()
 
         @make_tf_callable(self.get_session())
-        def do_update():
-            # update_target_fn will be called periodically to copy Q network to
-            # target Q network
+        def update_target_fn(tau):
+            tau = tf.convert_to_tensor(tau, dtype=tf.float32)
             update_target_expr = []
-            assert len(trainable_q_func_vars) == len(trainable_target_q_func_vars), (
-                trainable_q_func_vars,
-                trainable_target_q_func_vars,
+            assert len(model_vars) == len(target_model_vars), (
+                model_vars,
+                target_model_vars,
             )
-            for var, var_target in zip(
-                trainable_q_func_vars, trainable_target_q_func_vars
-            ):
-                update_target_expr.append(var_target.assign(var))
+            for var, var_target in zip(model_vars, target_model_vars):
+                update_target_expr.append(
+                    var_target.assign(tau * var + (1.0 - tau) * var_target)
+                )
                 logger.debug("Update target op {}".format(var_target))
             return tf.group(*update_target_expr)
 
-        self.update_target = do_update
+        # Hard initial update.
+        self._do_update = update_target_fn
+        # TODO: The previous SAC implementation does an update(1.0) here.
+        # If this is changed to tau ~= 1.0 the sac_loss_function test fails. Why?
+        # Also the test is not very maintainable, we need to change that unittest
+        # anyway.
+        # TODO: The unittest on learning_tests_stateless_cartpole_r2d2_fake_gpus fails
+        # on tf1 when we call update_target() here. The reason is unkown. This is a
+        # workaround that will be removed once we move to the new policy stack.
+        if update_target_in_init:
```

Comment:
maybe you can try moving hard initial update to be after the f1.global_variables_initializer() call here:
https://github.com/ray-project/ray/blob/master/rllib/policy/dynamic_tf_policy.py#L474
https://github.com/ray-project/ray/blob/master/rllib/policy/dynamic_tf_policy_v2.py#L685
this will make sure the entire graph is initialized.
basically TargetNetworkMixin.__init__() happens before the loss and graph gets properly initialized.

Reply:
This is sth that will be removed in the overhaul anyway. It doesn't make sense for me to spend more time finding the elegant solution here.